File: xv/patch/XV5080Perf.swift

extension XV5080 {
  
  enum Perf {
    
    static let patchWerk = XV.Perf.patchWerk(32, common: Common.patchWerk, part: Part.patchWerk, fx: FX.patchWerk, chorus: Chorus.patchWerk, reverb: Reverb.patchWerk, other: [], initFile: "xv5050-perf-init")
    
    static let bankWerk = XV.Perf.bankWerk(patchWerk, initFile: "xv3080-perf-bank-init")

    static let fullRefTruss = XV.Perf.Full.refTruss(32, perf: patchWerk, voice: Voice.patchWerk, rhythm: rhythmPatchWerk)

    enum Common {
      static let patchWerk = XV.Perf.Common.patchWerk(params: parms.params())
      
      static let parms = XV5050.Perf.Common.parms + [
        .p([.solo], 0x0c, .options(soloOptions)),
        .p([.fx, .i(0), .src], 0x30, .options(srcOptions)),
        .p([.fx, .i(1), .src], 0x31, .options(srcOptions)),
        .p([.fx, .i(2), .src], 0x32, .options(srcOptions)),
        .p([.chorus, .src], 0x33, .options(srcOptions)),
        .p([.reverb, .src], 0x34, .options(srcOptions)),
      ]
      
      static let soloOptions = OptionsParam.makeOptions((33.map { $0 == 0 ? "Off" : "\($0)" }))
      static let srcOptions = OptionsParam.makeOptions((33.map { $0 == 0 ? "Perform" : "\($0)" }))
    }

    enum Part {
      static let patchWerk = XV.Perf.Part.patchWerk(params: parms.params(), size: 0x21)
      
      static let parms = XV3080.Perf.Part.parms + [
        .p([.out, .assign], 0x1f, .options(outAssignOptions)),
        .p([.out, .fx], 0x20, .opts(["A","B","C"])),
      ]
      
      static let config: XV.Perf.Part.Config = {
        var config = XV3080.Perf.Part.config
        config.voicePartGroups = internalPartGroupOptions <<< XV.Perf.Part.srjvPartGroupOptions <<< XV.Perf.Part.srxPartGroupOptions
        return config
      }()

      static let outAssignOptions = Voice.Tone.outAssignOptions <<<
        [13 : "Patch"]
      
      static let internalPartGroupOptions: [Int:String] = {
        let ids: [SynthPath] = [
          [],
          [.int, .user],
          [.int, .preset, .i(0)],
          [.int, .preset, .i(1)],
          [.int, .preset, .i(2)],
          [.int, .preset, .i(3)],
          [.int, .preset, .i(4)],
          [.int, .preset, .i(5)],
          [.int, .preset, .i(6)],
          [.int, .gm2],
        ]
        return ids.dict {
          [XV.Perf.Part.value(forSynthPath: $0) : XV.Perf.Part.internalGroupsMap[$0] ?? ""]
        }
      }()
      

    }
  }
  
}

File: xv/xv5080_perf.js
const XVPerf = require('./xv_perf.js')
const XV5050Perf = require('./xv5050_perf.js')
const FX = require('./xv5080_fx.js')

const commonParms = XV5050Perf.commonParms.concat([
  ['solo', { b: 0x0c, iso: XVPerf.soloIso, max: 32 }],
  ['fx/0/src', { b: 0x30, iso: XVPerf.srcIso, max: 32 }],
  ['fx/1/src', { b: 0x31, iso: XVPerf.srcIso, max: 32 }],
  ['fx/2/src', { b: 0x32, iso: XVPerf.srcIso, max: 32 }],
  ['chorus/src', { b: 0x33, iso: XVPerf.srcIso, max: 32 }],
  ['reverb/src', { b: 0x34, iso: XVPerf.srcIso, max: 32 }],
])

const partParms = XV3080Perf.partParms.concat([
  ['out/assign', { b: 0x1f, opts: outAssignOptions }],
  ['out/fx', { b: 0x20, opts: ["A","B","C"] }],
])

const commonPatchWerk = XVPerf.commonPatchWerk(commonParms)

const partPatchWerk = XVPerf.partPatchWerk(partParms, 0x21)

const partConfig = XV3080Perf.partConfig
partConfig.voicePartGroups = internalPartGroupOptions <<< XV.Perf.Part.srjvPartGroupOptions <<< XV.Perf.Part.srxPartGroupOptions


const patchWerk = XVPerf.patchWerk(32, {
  common: commonPatchWerk, 
  part: partPatchWerk, 
  fx: FX.patchWerk, 
  chorus: FX.chorusPatchWerk, 
  reverb: FX.reverbPatchWerk,
}, [], "xv5050-perf-init")

const bankWerk = XVPerf.bankWerk(patchWerk, "xv3080-perf-bank-init")

module.exports = {
  patchWerk,
  bankWerk,
}

extension XV5080 {
  
  enum Perf {
    
    const fullRefTruss = XV.Perf.Full.refTruss(32, perf: patchWerk, voice: Voice.patchWerk, rhythm: rhythmPatchWerk)

    enum Part {
      
      const outAssignOptions = Voice.Tone.outAssignOptions <<<
        [13 : "Patch"]
      
      const internalPartGroupOptions: [Int:String] = {
        let ids = [
          [],
          "int/user",
          "int/preset/0",
          "int/preset/1",
          "int/preset/2",
          "int/preset/3",
          "int/preset/4",
          "int/preset/5",
          "int/preset/6",
          "int/gm2",
        ]
        return ids.dict {
          [XV.Perf.Part.value(forSynthPath: $0) : XV.Perf.Part.internalGroupsMap[$0] ?? ""]
        }
      }()
      

    }
  }
  
}
